[GTK] Prefer EGL over X11 where available

ceyusa opened the pull request:
#### 5948e8b03d0ca94615e87db2d3435717e36a510c
<pre>
[GTK] Prefer EGL over X11 where available
<a href="https://bugs.webkit.org/show_bug.cgi?id=236163">https://bugs.webkit.org/show_bug.cgi?id=236163</a>

Reviewed by Žan Doberšek.

Patch Authored by Carlos Garcia Campos &lt;cgarcia@igalia.com&gt; and Arcady Goldmints-Orlov.

* Source/WebCore/platform/graphics/GLContext.cpp:
(WebCore::GLContext::createContextForWindow): Create context from shared context if available. Then
create EGL context first and GLX later.
(WebCore::GLContext::createSharingContext): Create EGL shared context first and GLX later, if
WEBKIT_FORCE_GLX environment variable is not set.
* Source/WebCore/platform/graphics/egl/GLContextEGL.cpp:
(WebCore::GLContextEGL::getEGLConfig): Select visual compatible with window one, if compatible
function is available.
(WebCore::GLContextEGL::createWindowContext): If platform is X11 get window attributes and
instantiates a function for compatible to later compare it with specified visual id.
* Source/WebCore/platform/graphics/egl/GLContextEGL.h:
* Source/WebCore/platform/graphics/gstreamer/PlatformDisplayGStreamer.cpp: Move and guard GstGL
headers.
(createGstGLDisplay): Use GstGLPlatform to select the display type.
(PlatformDisplay::tryEnsureGstGLContext const): Get GstGLPlatform from shared context.
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::openGLAPI): Simplify the API string creation.
(WebKit::WebKitProtocolHandler::handleGPU): Unify detection if EGL is used, before it assumed always
GLX if X11.
(WebKit::nativeInterface): Deleted.

Canonical link: <a href="https://commits.webkit.org/254751@main">https://commits.webkit.org/254751@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/068bc27c9b5b0f8cdeb278ec4bf5e3d1925bd00b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90123 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34669 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20740 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99440 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33161 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82455 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/94633 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95773 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76951 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26263 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81215 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80999 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69262 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/81739 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/34322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/15076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/76639 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16021 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26517 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3336 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/35907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/38989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/79225 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/37810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35105 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17360 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

TingPing reviewed (changes requested):
Tested this on Nvidia 515.65.01 + X11 and it still does not work.

All rendered content is not visible. No clear errors or logs.

TingPing reviewed (approved):
This now works for me. :smiley:

carlosgcampos commented on Source/WebCore/platform/graphics/egl/GLContextEGL.cpp:
```diff
@@ -161,9 +178,43 @@ bool GLContextEGL::getEGLConfig(EGLDisplay display, EGLConfig* config, EGLSurfac
 {
+    Function<bool(int)> checkCompatibleVisuals = nullptr;
+#if PLATFORM(X11)
+    XUniquePtr<XVisualInfo> visualInfo;
```
Aha, I see the bug now. I think we can just transfer the ownership to the Function instead.

carlosgcampos commented on Source/WebCore/platform/graphics/egl/GLContextEGL.cpp:
```diff
@@ -161,9 +178,43 @@ bool GLContextEGL::getEGLConfig(EGLDisplay display, EGLConfig* config, EGLSurfac
+            return XUniquePtr<XVisualInfo>(XGetVisualInfo(x11Display, VisualIDMask, &templateVisualInfo, &visualInfoCount));
+        };
+
+        visualInfo = visualInfoForID(XVisualIDFromVisual(attributes.visual));
```
Keep the declaration here.

carlosgcampos commented on Source/WebCore/platform/graphics/egl/GLContextEGL.cpp:
```diff
@@ -161,9 +178,43 @@ bool GLContextEGL::getEGLConfig(EGLDisplay display, EGLConfig* config, EGLSurfac
+        if (!visualInfo)
+            return nullptr;
+
+        checkCompatibleVisuals = [&visualInfo, &visualInfoForID](int configVisualID) {
```
And visualInfo = WTFMove(visualInfo) here. And we should do the same for the visualInfoFoID lambda, I guess. Because it's declared inside the if and passed to the Function as reference, but released when getEGLConfig is called.

ceyusa committed:
[GTK] Prefer EGL over X11 where available
https://bugs.webkit.org/show_bug.cgi?id=236163

Reviewed by Žan Doberšek.

Patch Authored by Carlos Garcia Campos <cgarcia@igalia.com> and Arcady Goldmints-Orlov.

* Source/WebCore/platform/graphics/GLContext.cpp:
(WebCore::GLContext::createContextForWindow): Create context from shared context if available. Then
create EGL context first and GLX later.
(WebCore::GLContext::createSharingContext): Create EGL shared context first and GLX later, if
WEBKIT_FORCE_GLX environment variable is not set.
* Source/WebCore/platform/graphics/egl/GLContextEGL.cpp:
(WebCore::GLContextEGL::getEGLConfig): Select visual compatible with window one, if compatible
function is available.
(WebCore::GLContextEGL::createWindowContext): If platform is X11 get window attributes and
instantiates a function for compatible to later compare it with specified visual id.
* Source/WebCore/platform/graphics/egl/GLContextEGL.h:
* Source/WebCore/platform/graphics/gstreamer/PlatformDisplayGStreamer.cpp: Move and guard GstGL
headers.
(createGstGLDisplay): Use GstGLPlatform to select the display type.
(PlatformDisplay::tryEnsureGstGLContext const): Get GstGLPlatform from shared context.
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::openGLAPI): Simplify the API string creation.
(WebKit::WebKitProtocolHandler::handleGPU): Unify detection if EGL is used, before it assumed always
GLX if X11.
(WebKit::nativeInterface): Deleted.

Canonical link: https://commits.webkit.org/254751@main
